### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by JohnMcPMS**
## Change
This change has several performance improvements based on profiling.

### Installed package caching
A global process cache is stored for the unfiltered install data.  It is very basic for updates right now to make it more reasonable to service, but future changes could improve things to examine the system data and only do delta updates.  It could also be saved to disk to improve the performance of future processes.

The cache itself is not used by any callers, but instead in-memory copies are handed out.  The cache sets up event listeners for the current data sources and waits for these events before it will attempt to update again.  A full cache hit requires only copying the memory database, which SQLite has an efficient method for.

The only performance gain to the update path that is provided by this implementation is to reuse the names of MSIX packages, but since this accounts for a huge amount of creating the index it results in ~50% time reduction in creating an updated installed source (without any MSIX changes).

### ICU regex caching
The regular expressions that we use were being compiled repeatedly.  Since this is a fixed set of expressions, they are all now cached and copies are used for actual operation.

### Version lookup (string)
The version lookup by string (such as `winget show FOO -v VER`) was doing a full table scan.  This change shifts to pulling all of the version for the given identifier out and using `Version` object comparisons to find the proper result.  This has the side effect of fixing a somewhat annoying user interaction where trailing 0s in a version were still required to find from the command line.  One can now provide the smaller version string and still find the expected package version (eg `1.2` will now find version `1.2.0`).

### Version lookup (available version)
The round trip from retrieving the available version information to getting the manifest has been improved by including the manifest id value directly in the available version data.  The available version package object now caches the version string to manifest id data as well, meaning that any usage of a `PackageVersionKey` that was retrieved from `GetAvailableVersions` will directly return the package version object without querying the index.

### Let property lookups handle an unknown manifest id
The `GetProperty` code was enforcing that the manifest id was valid and then reusing existing code that did not handle a missing manifest.  Creating new methods that can handle the missing manifest allows us to remove the initial check.

## Validation
Added tests where appropriate, regression covers many other cases.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/3808)

### CODE REVIEW

**Comment by denelon** [comment]
Related to:
* https://github.com/microsoft/winget-cli/issues/3667

File: src/AppInstallerRepositoryCore/Microsoft/PredefinedInstalledSourceFactory.h [hunk 1]
```diff
@@ -34,6 +34,8 @@ namespace AppInstaller::Repository::Microsoft
             User,
             // Contains machine ARP and machine MSIX
             Machine,
+            // Same as None but creating the source reference causes the next Open to always update the cache
+            NoneWithForcedCacheUpdate,
```
Comment (florelis) [inline src/AppInstallerRepositoryCore/Microsoft/PredefinedInstalledSourceFactory.h:38]: Could we ever want to use anything else with forced cache update? If not, can you add a comment why not?

File: src/AppInstallerRepositoryCore/Microsoft/SQLiteIndexSource.cpp [hunk 2]
```diff
@@ -288,14 +288,34 @@ namespace AppInstaller::Repository::Microsoft
...
+                }
+
+                std::vector<SQLiteIndex::VersionKey> versions = source->GetIndex().GetVersionKeysById(m_idId);
+
+                for (const auto& vk : versions)
+                {
+                    std::string version = vk.VersionAndChannel.GetVersion().ToString();
+                    std::string channel = vk.VersionAndChannel.GetChannel().ToString();
+                    m_availableVersionKeys.emplace_back(source->GetIdentifier(), version, channel);
+                    m_availableVersionKeysMap.emplace(MapKey{ std::move(version), std::move(channel) }, vk.ManifestId);
```
Comment (florelis) [inline src/AppInstallerRepositoryCore/Microsoft/SQLiteIndexSource.cpp:315]: Why does this one not take the source ID?

File: src/AppInstallerCLIE2ETests/Interop/InstallInterop.cs [hunk 3]
```diff
@@ -657,5 +657,73 @@ public void GetApplicableInstaller()
...
+        /// Install exe and verify that we can find it as installed after.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous unit test.</returns>
+        [Test]
+        public async Task InstallExe_VerifyInstalledCatalog()
+        {
+            var installedCatalogReference = this.packageManager.GetLocalPackageCatalog(LocalPackageCatalog.InstalledPackages);
+
+            // Ensure package is not installed
+            string targetPackageProductCode = "{A499DD5E-8DC5-4AD2-911A-BCD0263295E9}";
```
Comment (florelis) [inline src/AppInstallerCLIE2ETests/Interop/InstallInterop.cs:671]: Nit: I had to look up what this GUID was. Can this be a constant somewhere else?

File: src/AppInstallerCLITests/TestHooks.h [hunk 4]
```diff
@@ -39,6 +39,9 @@ namespace AppInstaller
     namespace Repository::Microsoft
     {
         void TestHook_SetPinningIndex_Override(std::optional<std::filesystem::path>&& indexPath);
+
+        using GetARPKeyFunc = Registry::Key(*)(void*, Manifest::ScopeEnum, Utility::Architecture);
```
Comment (florelis) [inline src/AppInstallerCLITests/TestHooks.h:43]: Nit: Could we use `std::function` here?

File: src/AppInstallerCLITests/PredefinedInstalledSource.cpp [hunk 5]
```diff
@@ -400,3 +402,89 @@ TEST_CASE("PredefinedInstalledSource_Search", "[installed][list]")
...
+        GetDatabaseIdentifier(source1)
+        ==
+        GetDatabaseIdentifier(source2)
+    );
+
+    // Get all packages
+    auto result1 = source1->Search({});
+    REQUIRE(!result1.Matches.empty());
+
+    for (const auto& match : result1.Matches)
```
Comment (florelis) [inline src/AppInstallerCLITests/PredefinedInstalledSource.cpp:427]: Nit: Could you add a comment about what this is loop is doing?

Comment (florelis) [inline src/AppInstallerCLITests/PredefinedInstalledSource.cpp:427]: Could/should we check that there are no packages in result2 that aren't in result1?

File: src/AppInstallerCLITests/ArpHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    ARPHelper helper;
+
+    wil::unique_event callbackEvent;
+    callbackEvent.create();
+
+    ScopeEnum scopeCallback = ScopeEnum::Unknown;
+    Architecture architectureCallback = Architecture::Unknown;
+
+    ScopeEnum scopeTarget = ScopeEnum::User;
+    Architecture architectureTarget = Architecture::X64;
```
Comment (florelis) [inline src/AppInstallerCLITests/ArpHelper.cpp:26]: Will this test run on x86?

**Review by florelis — DISMISSED** [review]

File: azure-pipelines.yml [hunk 7]
```diff
@@ -295,18 +297,18 @@ jobs:
...
+  #   inputs:
+  #     testRunTitle: Microsoft.Management.Configuration.UnitTests (InProc)
+  #     testSelector: 'testAssemblies'
+  #     testAssemblyVer2: '**\Microsoft.Management.Configuration.UnitTests.dll'
+  #     searchFolder: '$(buildOutDir)\Microsoft.Management.Configuration.UnitTests'
+  #     codeCoverageEnabled: true
+  #     platform: '$(buildPlatform)'
+  #     configuration: '$(BuildConfiguration)'
+  #     diagnosticsEnabled: true
+  #   condition: succeededOrFailed()
```
Comment (florelis) [inline azure-pipelines.yml:311]:
```suggestion
  - task: VSTest@2
    displayName: 'Run tests: Microsoft.Management.Configuration.UnitTests (InProc)'
    inputs:
      testRunTitle: Microsoft.Management.Configuration.UnitTests (InProc)
      testSelector: 'testAssemblies'
      testAssemblyVer2: '**\Microsoft.Management.Configuration.UnitTests.dll'
      searchFolder: '$(buildOutDir)\Microsoft.Management.Configuration.UnitTests'
      codeCoverageEnabled: true
      platform: '$(buildPlatform)'
      configuration: '$(BuildConfiguration)'
      diagnosticsEnabled: true
    enabled: false
    condition: succeededOrFailed()

```

File: src/AppInstallerRepositoryCore/Microsoft/SQLiteIndexSource.cpp (hunk 2 above)
Comment (JohnMcPMS) [reply src/AppInstallerRepositoryCore/Microsoft/SQLiteIndexSource.cpp:315]: It isn't needed for the mapping.

**Review by florelis — APPROVED** [review]

**Review by yao-msft — APPROVED** [review]

File: src/AppInstallerSharedLib/Public/winget/SQLiteStorageBase.h [hunk 8]
```diff
@@ -25,6 +25,9 @@ namespace AppInstaller::SQLite
         // Gets the last write time for the database.
         std::chrono::system_clock::time_point GetLastWriteTime();
 
+        // Gets the identifier written to the database when it was created.
+        std::string GetDatabaseIdentifier();
```
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply src/AppInstallerSharedLib/Public/winget/SQLiteStorageBase.h:29]: I added it exclusively for test purposes at this time.  But it seemed like an interesting enough idea to just leave it in for all databases.

**Review by yao-msft — APPROVED** [review]

**Comment by Girofox** [comment]
A possible trigger for debugging is manually starting backup in **Windows Backup** app, this instantly causes _WindowsPackageManagerServer.exe_ running at around 25 % cpu usage for some time. 

The log file in `%LocalAppData%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\DiagOutputDir` grows very quickly to more than 1.0 Mb.

**Pull request merged**